File: code-monkey-ts/tools/ca/CADependencyGraphTool.ts
import { z } from 'zod';
import { AsyncCallbackManagerForToolRun } from '@langchain/core/callbacks/manager';
import { tool } from '@langchain/core/tools';
import { CATool } from './CATool';
import { resolveFilePath, resolveModulePath, getModuleName } from '../deps';
import { instrument } from '../../instrumentation';
import { createGraph } from 'ngraph.graph';

const CADependencyGraphToolInputSchema = z.object({
  files: z.array(z.string()).optional().describe("List of relative file paths to analyze"),
  modules: z.array(z.string()).optional().describe("List of modules to analyze")
}).refine(data => data.files != null || data.modules != null, {
  message: "Either 'files' or 'modules' must be provided"
});

type CADependencyGraphToolInput = z.infer<typeof CADependencyGraphToolInputSchema>;

export class CADependencyGraphTool extends CATool {
  name = "ca_generate_dependency_graph";
  description = "Generate a dependency graph for Python files";
  argsSchema = CADependencyGraphToolInputSchema;

  constructor() {
    super();
  }

  @instrument("Tool._run", ["files", "modules"], { attributes: { tool: "CADependencyGraphTool" } })
  async _run(
    input: CADependencyGraphToolInput,
    runManager?: AsyncCallbackManagerForToolRun,
  ): Promise<string> {
    const { files = [], modules = [] } = input;

    const allFiles = [
      ...files.map(resolveFilePath),
      ...modules.map(resolveModulePath)
    ];

    const graph = createGraph();

    for (const filePath of allFiles) {
      const moduleName = getModuleName(filePath);
      const imports = await this.parser.getImports(filePath);

      for (const imp of imports) {
        graph.addLink(moduleName, imp);
      }
    }

    return JSON.stringify(graph, null, 1);
  }
}

export const caDependencyGraphTool = tool({
  name: "ca_generate_dependency_graph",
  description: "Generate a dependency graph for Python files",
  schema: CADependencyGraphToolInputSchema,
  func: async (input: CADependencyGraphToolInput, runManager?: AsyncCallbackManagerForToolRun) => {
    const tool = new CADependencyGraphTool();
    return tool._run(input, runManager);
  }
});

File: code-monkey-py/src/tools/ca/ca_dependency_graph_tool.py
import json
from pydantic import BaseModel, Field
from typing import Type, List, Optional, Any
from langchain_core.callbacks import (
    AsyncCallbackManagerForToolRun,
)
import networkx as nx
from .ca_tool import CATool
from deps import resolve_file_path, resolve_module_path, get_module_name
from instrumentation import instrument


class CADependencyGraphToolInput(BaseModel):
    files: List[str] = Field(None, description="List of relative file paths to analyze")
    modules: List[str] = Field(None, description="List of modules to analyze")

    # not sure how to encode required rules at all (I think everything defaults to required), and certainly
    # don't know how to encode this one:
    #
    # "anyOf": [{"required": ["files"]}, {"required": ["modules"]}]


class CADependencyGraphTool(CATool):
    """Generate a dependency graph for Python files"""

    name = "ca_generate_dependency_graph"
    description = "Generate a dependency graph for Python files"
    args_schema: Type[BaseModel] = CADependencyGraphToolInput

    def __init__(self, **kwargs: Any) -> None:
        super().__init__(**kwargs)

    @instrument(
        "Tool._run", ["files", "modules"], attributes={"tool": "CADependencyGraphTool"}
    )
    def _run(
        self,
        files: List[str] | None,
        modules: List[str] | None,
        run_manager: Optional[AsyncCallbackManagerForToolRun],
    ) -> str:
        if files is None:
            files = []
        if modules is None:
            modules = []

        all_files = [resolve_file_path(f) for f in files] + [
            resolve_module_path(m) for m in modules
        ]

        graph = nx.DiGraph()

        for file_path in all_files:
            module_name = get_module_name(file_path)
            imports = self.parser.get_imports(file_path)

            for imp in imports:
                graph.add_edge(module_name, imp)

        return json.dumps(nx.node_link_data(graph), indent=1)
